psi4/src/psi4/lib3index/df_helper.cc

```diff
@@ -2183,51 +2447,58 @@ void DF_Helper::compute_J(std::vector<SharedMatrix> D, std::vector<SharedMatrix>
             rank = omp_get_thread_num();
             #endif
 
-            for (size_t m = 0, sp_count = -1; m < nao_; m++) {
+            for (size_t m = 0, sp_count = -1; m < nao; m++) {
                 if (schwarz_fun_mask_[k * nao + m]) {
                     sp_count++;
-                    C_DCOPY(1, &Dp[nao*k+m], 1, &D_buffers[rank][sp_count], 1);
+                    D_buffers[rank][sp_count] = Dp[nao*k+m];
                 }
             }
             // (Qm)(m) -> (Q)
-            C_DGEMV('N', block_size, sp_size, 1.0, &Mp[jump], sp_size, &D_buffers[rank][0], 1, 1.0, &T1p[rank*naux], 1);
+            C_DGEMV('N', block_size, sp_size, 1.0, &Mp[jump], sp_size, &D_buffers[rank][0], 1, 1.0, &T1p[rank * align_size], 1);
+        }
+        timer_off("DF_Helper~JK        - Jfirst");
+        timer_on("DF_Helper~JK       - Jreduce");
+        // reduce
+        for(size_t k=1; k<nthreads_; k++){
+            for(size_t l=0; l < align_size; l++)
+                T1p[l] += T1p[k * align_size + l];
+        }
+        timer_off("DF_Helper~JK       - Jreduce");
+        timer_on("DF_Helper~JK       - Jcomple");
+        
+        // align cache to reduce false sharing
+        align_size = (nao % 8 ? nao + (8 - nao % 8) : nao);
+        
+        // complete pruned J
+        #pragma omp parallel for schedule(guided) num_threads(nthreads_)
+        for(size_t k=0, count=0; k<nao; k++){
+            size_t sp_size = small_skips_[k];
+            size_t jump = (AO_core_ ? big_skips_[k] + bcount * sp_size : (big_skips_[k] * block_size) / naux);
+            C_DGEMV('T', block_size, sp_size, 1.0, &Mp[jump], sp_size, T1p, 1, 0.0, &T2p[k*align_size], 1);
```

Comment:
I would mark this as a point to revisit, we could still save ~50% on J due to symmetry. It wouldnt impact the total walltime that much, but could help the total efficiency marks.